core/src/main/scala/com/metamx/tranquility/tranquilizer/Tranquilizer.scala

```diff
@@ -0,0 +1,305 @@
+              pendingBatches >= maxPendingBatches ||
+              (lingerMillis > 0 && System.currentTimeMillis() < bufferStartMillis + lingerMillis)
+          ) {
+            if (lingerMillis == 0 || bufferStartMillis == 0) {
+              lock.wait()
+            } else {
+              lock.wait(math.max(1, bufferStartMillis + lingerMillis - System.currentTimeMillis()))
+            }
+          }
+          if (log.isDebugEnabled) {
+            log.debug(s"Buffer with ${buffer.size} messages has lingered too long, sending.")
+          }
+
+          swap()
+        }
+
+        sendBuffer(exBuffer)
+      }
+    }
+    catch {
+      case e: InterruptedException =>
+        log.debug("Interrupted, exiting thread.")
+    }
+  } withEffect { t =>
+    t.setDaemon(true)
+    t.setName(s"Tranquilizer-BackgroundSend-[$beam]")
+  }
+
+  /**
+    * Start the tranquilizer. Must be called before any calls to "send".
+    */
+  @LifecycleStart
+  def start(): Unit = {
+    started = true
+    sendThread.start()
+  }
+
+  /**
+    * Sends a message. Generally asynchronous, but will block if the internal buffer is full.
+    *
+    * The future may contain a Unit, in which case the message was successfully sent. Or it may contain an exception,
+    * in which case the message may or may not have been successfully sent. One specific exception to look out for is
+    * MessageDroppedException, which means that a message was dropped due to unrecoverable reasons. With Druid this
+    * can be caused by message timestamps being outside the configured windowPeriod.
+    *
+    * NB: This can actually be wrong: if *any* message in a batch of messages is dropped, then all messages in that
+    * batch will be marked as dropped (even if they actually made it out). This is a limitation of how Tranquility
+    * tracks which messages were and were not dropped. This could be improved once
+    * https://github.com/druid-io/tranquility/issues/56 is addressed.
+    */
+  def send(message: MessageType): Future[Unit] = {
+    requireStarted()
+    if (log.isTraceEnabled) {
+      log.trace(s"Sending message: $message")
+    }
+    val holder = MessageHolder(message, Promise())
+
+    val (exBuffer, future) = lock.synchronized {
+      while (buffer.size >= maxBatchSize - 1 && pendingBatches >= maxPendingBatches) {
+        if (log.isDebugEnabled) {
+          log.debug(
+            s"Buffer size[${buffer.size}] >= maxBatchSize[$maxBatchSize] - 1 and " +
+              s"pendingBatches[$pendingBatches] >= maxPendingBatches[$maxPendingBatches], waiting..."
+          )
+        }
+        lock.wait()
+      }
+
+      if (buffer.isEmpty) {
+        bufferStartMillis = System.currentTimeMillis()
+        lock.notifyAll()
+      }
+
+      buffer += holder
+
+      val _exBuffer: Option[Buffer[MessageHolder]] = if (buffer.size == maxBatchSize) {
+        Some(swap())
+      } else if (lingerMillis == 0 && pendingBatches < maxPendingBatches) {
```

Comment:
I don't think we can, because at this point buffer.size < maxBatchSize, which means pendingBatches could potentially be >= maxPendingBatches. And we have to check for that, as you are not allowed to swap unless there is room (pendingBatches < maxPendingBatches).